docs/source/cpp/dataset.rst

```diff
@@ -0,0 +1,381 @@
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. default-domain:: cpp
+.. highlight:: cpp
+
+================
+Tabular Datasets
+================
+
+.. seealso::
+   :doc:`Dataset API reference <api/dataset>`
+
+.. warning::
+
+    The ``arrow::dataset`` namespace is experimental, and a stable API
+    is not yet guaranteed.
+
+The Arrow Datasets library provides functionality to efficiently work with
+tabular, potentially larger than memory and multi-file datasets:
+
+* A unified interface for different sources: supporting different sources and
+  file formats (Parquet, Feather files) and different file systems (local,
+  cloud).
+* Discovery of sources (crawling directories, handle directory-based partitioned
+  datasets, basic schema normalization, ..)
+* Optimized reading with predicate pushdown (filtering rows), projection
+  (selecting columns), parallel reading, or fine-grained managing of tasks.
+
+Currently, only Parquet, Feather / Arrow IPC, and CSV files are supported. The
+goal is to expand this in the future to other file formats and data sources
+(e.g.  database connections).
+
+Reading Datasets
+----------------
+
+For the examples below, let's create a small dataset consisting
+of a directory with two parquet files:
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 50-85
+   :linenos:
+   :lineno-match:
+
+(See the full example at bottom: :ref:`cpp-dataset-full-example`.)
+
+Dataset discovery
+~~~~~~~~~~~~~~~~~
+
+A :class:`arrow::dataset::Dataset` object can be created using the various
+:class:`arrow::dataset::DatasetFactory` objects. Here, we'll use the
+:class:`arrow::dataset::FileSystemDatasetFactory`, which can create a dataset
+given the base directory path:
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 151-170
+   :emphasize-lines: 6-11
+   :linenos:
+   :lineno-match:
+
+Here, we're also passing the filesystem to use and the file format to
+read. This lets us choose between things like reading local files or files in
+Amazon S3, or between Parquet and CSV.
+
+In addition to a base directory path, we can list file paths manually.
+
+Creating a :class:`arrow::dataset::Dataset` in this way loads nothing into
+memory; it only crawls the directory to find all the files
+(:func:`arrow::dataset::FileSystemDataset::files`):
+
+.. code-block:: cpp
+
+   for (const auto& filename : dataset->files()) {
```

Comment:
It seems a bit odd that here we iterate through and print `dataset->files()` when in an earlier example we were iterating `dataset->GetFragments()`.  Although I suppose I understand it is because we are working with the more specialized `FileSystemDataset` and not a plain old `Dataset`.